File: data-structure/js/binary-search.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderedArray = void 0;
var OrderedArray = /** @class */ (function () {
    function OrderedArray(length) {
        this.lastPos = -1;
        this.capacity = length;
        this.values = new Array(length);
    }
    OrderedArray.prototype.print = function () {
        if (this.lastPos === -1)
            return console.log("The array is empty");
        for (var index in this.values) {
            console.log("".concat(index, " -- ").concat(this.values[index]));
        }
        console.log(this.values);
    };
    OrderedArray.prototype.insert = function (value) {
        if (this.lastPos + 1 === this.values.length)
            return console.log("The array is filled");
        var position = 0;
        for (var i = position; i < this.lastPos + 1; i++) {
            position = i;
            if (this.values[i] > value)
                break;
            if (this.lastPos === i)
                position = i + 1;
        }
        var x = this.lastPos;
        while (x >= position) {
            this.values[x + 1] = this.values[x];
            x--;
        }
        this.values[position] = value;
        this.lastPos++;
    };
    OrderedArray.prototype.bsearch = function (value) {
        if (this.lastPos === -1)
            return -1;
        var min = 0;
        var max = this.lastPos;
        while (min <= max) {
            var mid = Math.ceil((min + max) / 2);
            if (this.values[mid] === value)
                return mid;
            if (this.values[mid] > value)
                max = mid - 1;
            if (this.values[mid] < value)
                min = mid + 1;
        }
        return -1;
    };
    return OrderedArray;
}());
exports.OrderedArray = OrderedArray;
var orderedArray = new OrderedArray(5);
orderedArray.print();
orderedArray.insert(5);
orderedArray.insert(3);
orderedArray.insert(6);
orderedArray.insert(1);
orderedArray.insert(8);
orderedArray.print();
console.log(orderedArray.bsearch(3));
console.log(orderedArray.bsearch(8));
console.log(orderedArray.bsearch(5));
console.log(orderedArray.bsearch(1));

File: data-structure/python/arrays/binary-search.py
import numpy as np

class OrderedArray:
    def __init__(self, capacity):
        self.capacity = capacity
        self.last_pos = -1
        self.values = np.empty(self.capacity, dtype=int)

    def print(self):
        if (self.last_pos == -1):
            return print('Array empty')
        
        for i in range(self.last_pos + 1):
            print(f"{i} -- {self.values[i]}")
    
    def insert(self, value):
        if self.last_pos + 1 == self.capacity:
            return print('Array already filled')
        
        position = 0
        for i in range(self.last_pos + 1):
            position = i
            if self.values[i] > value:
                break
            if i == self.last_pos:
                position = i + 1
        
        aux = self.last_pos

        while aux >= position:
            self.values[aux + 1] = self.values[aux]
            aux -= 1

        self.values[position] = value
        self.last_pos += 1
    
    def bsearch(self, value):
        if self.last_pos == -1:
            return -1
        
        min = 0
        max = self.last_pos

        while min <= max:
            mid = int((min + max) / 2 )
            if self.values[mid] == value:
                return mid
            if value > self.values[mid]:
                min = mid + 1
            if value < self.values[mid]:
                max = mid - 1
        
        return -1
            


        
ordered = OrderedArray(5)
ordered.print()
ordered.insert(5)
ordered.insert(3)
ordered.insert(6)
ordered.insert(8)
ordered.insert(1)
ordered.print()
print(ordered.bsearch(3))
print(ordered.bsearch(6))
print(ordered.bsearch(1))